File: go/1. Sleep/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func routine(millis int, ch chan<- float64) {
	var n float64 = 0
	for i := 0; i < 50_000; i++ {
		if n > 1000 {
			n /= 850
		}
		n *= 4
	}
	time.Sleep(time.Duration(millis) * time.Millisecond)
	ch <- n
}

func lcg(n, maxValue int) []int {
	timers := make([]int, n)
	timers[0] = 42

	m := 0x7FFFFFFF
	a := 48271
	c := 0
	for i := 1; i < n; i++ {
		timers[i] = (timers[i-1]*a + c) % m
		timers[i-1] = timers[i-1] % maxValue
	}
	timers[n-1] = timers[n-1] % maxValue

	return timers
}

func main() {
	n, err := strconv.Atoi(os.Args[1])
	if err != nil {
		panic(err)
	}

	timers := lcg(n, 300)

	start := time.Now()
	var chs []chan float64
	for i := 0; i < n; i++ {
		ch := make(chan float64, 1)
		chs = append(chs, ch)
		go routine(timers[i], ch)
	}

	list := make([]float64, n)
	for i := 0; i < n; i++ {
		list[i] = <-chs[i]
	}

	_ = len(list)

	fmt.Println(time.Since(start).Microseconds())
}

File: rust/1. Sleep/src/main.rs
use std::time::{Duration, Instant};
use tokio::task::JoinHandle;
use tokio::time::sleep;

async fn routine(millis: u64) -> f64 {
    let mut n: f64 = 0.;
    for i in 1..50_000 {
        if n > 1000. {
            n /= 850.;
        }
        n *= 4.;
    }
    sleep(Duration::from_millis(millis)).await;
    n
}

fn lcg(n: usize, max_value: i128) -> Vec<i32> {
    let mut list: Vec<i128> = Vec::with_capacity(n);
    list.push(42);

    let m = 0x7FFFFFFF;
    let a = 48271;
    let c = 0;
    for i in 1..n {
        list.push((list[i - 1] * a + c) % m);
        list[i - 1] = list[i - 1] % max_value;
    }
    list[n - 1] = list[n - 1] % max_value;

    list.iter().map(|it| *it as i32).collect()
}

#[tokio::main]
async fn main() {
    let n = std::env::args().nth(1).unwrap().parse().unwrap();
    let generated = lcg(n, 300);
    let timers = generated.iter();

    let start = Instant::now();

    let handles: Vec<JoinHandle<_>> = timers.map(|i| {
        tokio::spawn(routine(*i as u64))
    }).collect();

    let mut list = Vec::with_capacity(n);
    for handle in handles {
        list.push(handle.await.unwrap());
    }
    _ = list.len();

    println!("{:?}", start.elapsed().as_micros());
}
